README.md
# fview-flex

So IjzerenHein's [famous-flex](https://github.com/IjzerenHein/famous-flex)
is pretty friggin' awesome, and without a doubt the best community component for
Famo.us by a substantial distance.  This is the start of a
[famous-views](http://famous-views.meteor.com/) wrapper for it (for Meteor).

**VERY EARLY RELEASE!!!  Feel free to play around but don't use this for
anything serious.**  Works with both `mjn:famous` and  `raix:famono` (since
v0.0.2).

[Live demo](http://fview-flex.meteor.com/)

Too add to your project, just `meteor add gadicohen:fview-flex`.  To clone from
github, see **Development** below (in short, make sure to submodule init/update).

## Usage:

```jade
  +ContainerSurface perspective=500 overflow="hidden"
    +FlexScrollView layout="WheelLayout" direction="Y" layoutOptions=layoutOptions
      +famousEach surfaces
        +Surface style=style
          | #{_id}
```

Note: [Beware of overflow: hidden, it breaks z-translation quite badly](https://github.com/Famous/famous/issues/493) - if you don't need it, don't use it.

**Attributes**:

* All attributes are parsed using famous-views and given on init (i.e. you can give either a string or [fview string](http://famous-views.meteor.com/views/README) in the template, or use a template helper to give an actual value).  So you can use any option mentioned
in the famous-flex docs.
* `layout`, `direction`, `layoutOptions` are all reactive, see the
[demo source](https://github.com/gadicc/fview-flex/tree/master/demo/client)
for a good example.  
* If none of your layoutOptions are reactive, you can use straight JSON
directly from the template, e.g.
`layoutOptions='{ "itemSize": [50,90], "diameter": 500, "radialOpacity": 0 }'`.
* No problem to use a helper to return your own custom layout function to
`layout`.

You can also use `{{#FlexLayoutController}}` directly if you don't need
scrolling.

## Special behaviour

* If the FlexScrollView's immediate parent is a ContainerSurface, we'll set up
all the event piping for you.  Otherwise, by default, we'll still automatically
pipe all children to it, just like in regular ScrollView with famousEach.  But
generally you'll want to use a ContainerSurface to ensure smooth scrolling even
when going over "gaps" between the Surfaces.

## Progress

**famous-views** is developed in my spare time.  **fview-flex** is developed in
my spare time from famous-views :)  I'm not sure how quickly I'll get the full
list below implemented, however, I'll definitely prioritize according to demand.
[Open an issue](https://github.com/gadicc/fview-flex/issues) for feature /
prioritization requests, and I'll try get to these first, especially if there
are a lot of `+1`'s.

Basically you can use all the components of famous-flex right now, through js if not available for templates.

* **Standard Layouts** ([full docs](https://github.com/IjzerenHein/famous-flex#standard-layouts))
  * *--- Non-scrollable ---*
  * [x] GridLayout
  * [x] ProportionalLayout
  * [x] HeaderFooterLayout (compared to famo.us' HeaderFooterLayout, this could update the header/footer size dynamically)
  * [x] NavBarLayout
  * [x] TabBarLayout
  * *--- Scrollable ----*
  * [x] ListLayout ([x] stickyHeaders)
  * [x] CollectionLayout
  * [x] WheelLayout

* **LayoutHelpers(layout literals)**
  * [ ] LayoutDockHelper (add template support for dock literals)

* **Widgets**
  * [ ] DatePicker (add template support)
  * [ ] TabBar (add template support)

* **LayoutControllers**, **LayoutHelpers**, etc.

A DatePicker example(will add template support in the future :):
```
  var datePicker = new Flex.DatePicker({
    date: new Date(),
    perspective: 500,
    wheelLayout: {
      itemSize: 25,
      diameter: 100,
      radialOpacity: -0.5
    },
    createRenderables: {
      top: true,
      bottom: true
    },
    classes: ['transparent']
  });
  datePicker.setComponents([
    new Flex.DatePicker.Component.Month(),
    new Flex.DatePicker.Component.Day(),
    new Flex.DatePicker.Component.Year()
  ]);

  datePicker.on('datechange', function(event) {
    console.log('date-changed to: ' + event.date.toString());
  });
```

A HeaderFooterLayout example:
```
{{#FlexLayoutController layout='HeaderFooterLayout' layoutOptions=layoutOptions}}
  {{#Surface target="header" size="[undefined, 40]"}}
  {{/Surface}}
  {{#Surface target="content"}}
  {{/Surface}}
  {{#Surface target="footer" size="[undefined, 80]"}}
  {{/Surface}}
{{/FlexLayoutController}}
```

A LayoutDockHelper example:
```
The template:

<template name="dockExample">
{{#FlexLayoutController layout=layout dataSource=dataSource layoutOptions=layoutOptions}}
{{/FlexLayoutController}}
</template>

In JS:

Template.dockExample.helpers({
  'layout': function() {
    return {
      dock: [
        ['fill', 'background'],
        ['left', undefined, 8],
        ['top', undefined, 8],
        ['right', undefined, 8],
        ['bottom', undefined, 8],
        ['right', 'send', undefined, 1],
        ['fill', 'input', 1]
      ]
    };
  },
  'dataSource': function() {
    return {
      background: backgroundSurface,
      input: inputSurface,
      send:  sendSurface
    };
  },
  'layoutOptions': function() {
    return {
      margins: [0, 0, 0, 0]
    };
  }
});
```
I think future release can add support for layout literals(and other common layouts) like this:
```
{{#FlexLayoutController layout='HeaderFooterLayout/ListLayout/Dock/or whatever layout' margin="[10, 10, 10, 10]"}}
  {{#Surface target="left" size="[undefined, 100]"}}
    {{>templateA}}
  {{/Surface}}
  {{#Surface target="right" size="[undefined, 100]"}}
    {{>templateB}}
  {{/Surface}}
{{/FlexLayoutController}}
```

BTW, you can use RefreshLoader & AutosizeTextareaSurface like this:
```
var loader = new Flex.RefreshLoader(...);
```

## Development

If you plan to contribute, you probably want to clone your own fork.
See https://help.github.com/articles/using-pull-requests/ for more details.

```bash
$ git clone https://github.com/gadicc/fview-flex.git
$ cd fview-flex
$ git submodule init
$ git submodule update
```

The last two lines are necessary because we don't include all the original
fmaous-flex etc code in the repo, rather, we include them as submodules and
can update / checkout a particular commit whenever we want.  If you don't
you'll see errors like:

```bash
=> Errors while initializing project:

While building package gadicohen:fview-flex:
error: File not found: lib/famous-flex/src/LayoutUtility.js
error: File not found: lib/famous-flex/src/LayoutContext.js
```

etc.

History.md
## vNEXT

## v0.0.8

* TabBar: `fview.index()` -> `fview.selectedTab()`, and `fview.selectedTab(valueToSet)`
* TabBar: `selectedTab` attribute in template declaration, from a reactive helper

## v0.0.7

* Update to latest famous-flex, famous-autosizetextarea, famous-refresh-loader
* Famous-views wrapper for TabBar
* Extra Flex modules: LayoutNodeManager, LayoutController,
  AnimationController, TabBarController

## v0.0.6

* Fix for AutosizeTextareaSurface.prototype error (deps)

## v0.0.5

* add HeaderFooterLayout
* add LayoutDockHelper
* add DatePicker & TabBar
* add useful views RefreshLoader & AutosizeTextareaSurface
* bugfixes

A massive thanks to @ShawnOceanHu for his work on *all of the above*, including
sample code in the README.

## v0.0.4

* Bugfix: don't pipe events from parent ContainerSurface AND child surfaces,
  as the double piping breaks fluid flicking.  Instead, do one or the other,
  and also, change the way we pipe ContainerSurface to work better on Mac.

* Bump famous-flex, work correctly with now deprecated GridLayout.
  thanks @ShawnOceanHu.

## v0.0.3

* Bugfix: don't bail on `Cannot set property '_isDirty' of null` on removedAt
  if the containing view destroyed before defer time.

## v0.0.2

* famono support
* Demo: .noselect on surfaces makes it a tiny bit easier (but still pretty
  hard) to "flick" with mouse


lib/pre.js
Flex = window.Flex = {};

var modules = [
  'famous-flex/src/LayoutUtility',
  'famous-flex/src/LayoutContext',
  'famous-flex/src/LayoutNode',
  'famous-flex/src/LayoutNodeManager',
  'famous-flex/src/FlowLayoutNode',
  'famous-flex/src/helpers/LayoutDockHelper',
  'famous-flex/src/LayoutController',
  'famous-flex/src/AnimationController',

  'famous-flex/src/LayoutNodeManager',
  'famous-flex/src/LayoutController', 
  'famous-flex/src/ScrollController',
  'famous-flex/src/VirtualViewSequence',

  'famous-flex/src/layouts/CollectionLayout',
  'famous-flex/src/layouts/GridLayout',
  'famous-flex/src/layouts/ListLayout',
  'famous-flex/src/layouts/ProportionalLayout',
  'famous-flex/src/layouts/WheelLayout',
  'famous-flex/src/layouts/HeaderFooterLayout',
  'famous-flex/src/layouts/TabBarLayout',

  'famous-flex/src/widgets/DatePickerComponents',
  'famous-flex/src/widgets/DatePicker',
  'famous-flex/src/widgets/TabBar',
  'famous-flex/src/widgets/TabBarController',

  'famous-refresh-loader/RefreshLoader',
  'famous-autosizetextarea/AutosizeTextareaSurface',

  'famous-flex/src/FlexScrollView'
];

require = function(obj) {
  //console.log('r', obj);
  obj = obj.split('/');
  if (obj[0] === 'famous') {
    var s, ret = window;
    while (s=obj.shift())
      ret = ret[s];
    return ret;
  } else if (obj[obj.length-1].split('.')[0] === 'polyfill') {
  } else {
    var moduleName = obj[obj.length-1];
    var ret = Flex[moduleName];
    if (!ret)
      console.warn('require ' + moduleName + ' (not found) - needed by (probably) ' + currentDefine);
    return ret;
  }
}

var moduleCount = 0;
function functionName(fun) {
  var moduleName = modules[moduleCount++];
  return moduleName.split('/').pop();
}

var currentDefine = null;
define = function(func) {
  FView.ready(function() {
    //console.log('d', func);
    var exports = {}, module = {};
    currentDefine = func.toString().match(/module.exports = (\w+);/);
    if (currentDefine) currentDefine = currentDefine[1];
    func(require, exports, module);

    //if (typeof module.exports === 'function') {
      var moduleName = functionName(module.exports);
      Flex[moduleName] = module.exports;
      // console.log('d', moduleName);
    //} else {
    //  console.log('doh', typeof module.exports, module.exports);
    //}
  });
};

// imports
if (Package['raix:famono']) {
  FView.ready(function() {
    famous.surfaces.TextareaSurface;    // views/AutosizeTextareaSurface
    famous.transitions.Easing;          // widgets/TabBarController
  });
}

package.js
Package.describe({
  name: 'gadicohen:fview-flex',
  version: '0.0.9',
  summary: 'IjzerenHein\'s famous-flex for famous-views',
  git: 'https://github.com/gadicc/fview-flex.git'
});

Package.onUse(function(api) {
  // api.versionsFrom('1.0.3.1');
  api.use('reactive-dict@1.0.4', 'client');

  api.use('mjn:famous@0.3.5', 'client', { weak: true });
  api.use('raix:famono@0.9.27', 'client', { weak: true });

  api.use('gadicohen:famous-views@0.2.0', 'client');

  // custom require/define funcs
  api.addFiles('lib/pre.js', 'client');

  // ALWAYS copy this exactly into pre.js on update.  for now.
  var modules = [
    'famous-flex/src/LayoutUtility',
    'famous-flex/src/LayoutContext',
    'famous-flex/src/LayoutNode',
    'famous-flex/src/LayoutNodeManager',
    'famous-flex/src/FlowLayoutNode',
    'famous-flex/src/helpers/LayoutDockHelper',
    'famous-flex/src/LayoutController',
    'famous-flex/src/AnimationController',

    'famous-flex/src/LayoutNodeManager',
    'famous-flex/src/LayoutController', 
    'famous-flex/src/ScrollController',
    'famous-flex/src/VirtualViewSequence',

    'famous-flex/src/layouts/CollectionLayout',
    'famous-flex/src/layouts/GridLayout',
    'famous-flex/src/layouts/ListLayout',
    'famous-flex/src/layouts/ProportionalLayout',
    'famous-flex/src/layouts/WheelLayout',
    'famous-flex/src/layouts/HeaderFooterLayout',
    'famous-flex/src/layouts/TabBarLayout',

    'famous-flex/src/widgets/DatePickerComponents',
    'famous-flex/src/widgets/DatePicker',
    'famous-flex/src/widgets/TabBar',
    'famous-flex/src/widgets/TabBarController',

    'famous-refresh-loader/RefreshLoader',
    'famous-autosizetextarea/AutosizeTextareaSurface',

    'famous-flex/src/FlexScrollView'
  ];

  for (var i=0; i < modules.length; i++)
    api.addFiles('lib/' + modules[i] + '.js', 'client');

  // add the css files
  api.addFiles('lib/famous-flex/src/widgets/styles.css', 'client');

  // famous-views wrappers for famous-flex
  api.addFiles([
    'lib/FlexScrollView.js',
    'lib/TabBar.js'
  ], 'client');

  api.export('Flex', 'client');
});

Package.onTest(function(api) {
  api.use('tinytest');
  api.use('gadicohen:fview-flex');
});


lib/TabBar.js
var Map = function() {
  this.idToIdx = {};
  this.idxToId = {};
};
Map.prototype.add = function(idx, id) {
  this.idToIdx[id] = idx;
  this.idxToId[idx] = id;
};
Map.prototype.toId = function(idx) {
  return this.idxToId[idx] !== undefined ? this.idxToId[idx] : idx;
};
Map.prototype.toIdx = function(id) {
  return this.idToIdx[id] !== undefined ? this.idToIdx[id] : id;
};

FView.ready(function() {
  FView.registerView('TabBar', Flex.TabBar, {

    create: function(options) {
      var fview = this;

      fview.properties = new ReactiveDict();
      fview.selectedTab = function(value) {
        if (value)
          return this.properties.set('selectedTab', value);
        else
          return this.properties.get('selectedTab');
      };

      fview.items = [];
      fview.itemMap = new Map();

      if (options.class) {
        options.classes = options.class.split(' ');
        delete options.class;
      }

      if (!options.createRenderables)
        options.createRenderables = {};

      if (!options.createRenderables.item) {
        options.createRenderables.item = function(id, data) {
          // TODO https://github.com/IjzerenHein/famous-flex/blob/master/src/widgets/TabBar.js#L183

          // by default we'll accept full Surfaces / whatever
          return data.surface || data.view || data.node;
        }
      }

      var tabBar = new fview._view.constructor(options);

      tabBar.on('tabchange', function(event) {
        fview.properties.set('selectedTab', fview.itemMap.toId(event.index));
      });

      return tabBar;
    },

    add: function(child_fview, child_options) {
      var fview = this;

      if (child_options.tabId)
        fview.itemMap.add(fview.items.length, child_options.tabId);
      else if (child_options.id)
        fview.ItemMap.add(fview.items.length, child_options.id);

      var selectedTab = fview.selectedTab();

      // This won't be undefined if the app code initted it already, e.g. from router
      if (fview.items.length === 0 && selectedTab === undefined)
        fview.properties.set('selectedTab', child_options.tabId || child_options.id || 0);

      fview.items.push(child_fview);
      fview.view.setItems(fview.items);

      if (selectedTab === (child_options.tabId || fview.selectedTab || fview.items.length-1))
        fview.view.setSelectedItemIndex(fview.items.length-1);
    },

    attrUpdate: function(key, value, oldValue, data, firstTime) {
      var fview = this;

      if (key === 'selectedTab') {
        // On init, just the reactive var, setSelectedItemIndex called in add()
        if (firstTime)
          fview.properties.set('selectedTab', value);
        else
          fview.view.setSelectedItemIndex(fview.itemMap.toIdx(value));
      }
    }

  });
});

demo/client/demo.js
FView.ready(function() {
  // Famono: load famo.us shims and CSS
  famous.polyfills;
  famous.core.famous; // CSS
});

FlowRouter.route('/', {
  action: function() {
    FlowRouter.go('/FlexScrollView');
  }
});

//var getSession = function(sessionVar) { return Session.get('direction') }

Template.registerHelper('dstache', function() {
  return '{{';
});

Template.body.helpers({
  currentRouteName: function() {
    return FlowRouter.getRouteName();
  },
  layoutController: {
    flowOptions: {
      spring: {
        dampingRatio: 0.2,
        period: 500
      }
    }
  },
  createRenderables: {
    background: true,
    selectedItemOverlay: true
  },
  tabBarRender: function() {
    var fview = FView.from(this);    
    this.autorun(function(c) {
      FlowRouter.go(fview.selectedTab());
    });
  }
});

/*
Template.TabBarInitHack.onRendered(function() {
  // Since we're a non-fview template, the fview we get here is our parent's
  var fview = FView.from(this);

  this.autorun(function(c) {
    // only reactive dep
    var tabId = fview.selectedTab();

    // give precendence to route, not initial tab index
    if (c.firstRun)
      return;

    console.log(tabId, FlowRouter.current().route.name);
    if (tabId && tabId !== FlowRouter.current().route.name /* non-reactive *//*)
      FlowRouter.go(tabId);
  });
});
*/

lib/LayoutController.js
FView.registerView('FlexLayoutController', Flex.LayoutController, {

  create: function(options) {
    options = _.extend({
      layoutOptions: {
          itemSize: [100, 100],
          margins: [20, 20],
          spacing: [20, 20],
          justify: true,
          cells: [3,3],
          ratios: [0.2, 0.5, 0.3]
      },
      flow: true,    // smoothly animates renderables when changing the layout
      direction: undefined,   // 0 = X, 1 = Y, undefined = use default from selected layout-function
      autoPipeEvents: true,
      alignment: 1
    }, options);

    if (typeof options.layout === 'string') {
      if (Flex[options.layout])
        options.layout = Flex[options.layout];
      else
        throw new Error("[FlexLayoutController] No such Layout " + options.layout);
    }

    var node = new this._view.constructor(options);

    /*
     * If no dataSource is specified, expect a typical famousEach setup
     */
    if (!options.dataSource) {
      // Quick hack for init order
      node.sequenceFrom = function(source) {
        node.setDataSource(source);
      };

      this.reflow = _.debounce(function() {
        famous.core.Engine.defer(function() {
          node.reflowLayout();
        });
      }, 10);
    }

    return node;
  },

  famousCreatedPost: function() {
    this.pipeChildrenTo = this.parent.pipeChildrenTo ?
      [ this.view, this.parent.pipeChildrenTo[0] ] :
      [ this.view ];
  },

  attrUpdate: function(key, value, oldValue, data, firstTime) {

    if (key === 'layout') {
      if (typeof value === 'string') {
        if (Flex[value])
          return this.view.setLayout(Flex[value]);
        else
          throw new Error("[FlexLayoutController] No such Layout " + value);
      } else
        return this.setLayout(value);
    }

    if (key === 'direction')
      return this.view.setDirection(value);

    if (key === 'layoutOptions')
      return this.view.setLayoutOptions(value);

  },

  addedAt: function(id, item, index, _super, eachView) {
    _super(); this.reflow();
  },
  movedTo: function(id, doc, fromIndex, toIndex, _super, eachView) {
    _super(); this.reflow();
  },
  removedAt: function(doc, index, _super, eachView) {
    _super(); this.reflow();
  }
  
});

demo/client/FlexScrollView.js
FlowRouter.route('/FlexScrollView', { name: 'FlexScrollViewDemo' } );

// TODO, make session vars template state vars?

var layoutOptions = {
  GridLayout: [ 'cells', 'margins', 'spacing' ],
  ProportionalLayout: [ 'ratios' ],
  ListLayout: [ 'itemSize', 'margins', 'spacing' ],
  CollectionLayout: [ 'itemSize', 'justify', 'margins', 'spacing' ],
  // FullScreen: [ 'margins', 'spacing' ],
  WheelLayout: [ 'itemSize', 'diameter', 'radialOpacity' ]
};

Session.setDefault('cells', [4,4]);
Session.setDefault('margins', [10,10,10,10]);
Session.setDefault('ratios', [1,2,3,1]);
Session.setDefault('justify', [0,0]);
Session.setDefault('diameter', 500);
Session.setDefault('radialOpacity', 0);
// set in autorun below
// Session.setDefault('itemSize', [90,90]);
// Session.setDefault('spacing', [10,10]);

// Switch between scalar/array for itemSize when necessary w/ history
var itemSizeOld = [ [150,150], 70 ];
var spacingOld = [ [10,10], 2 ];
Tracker.autorun(function(c) {
  var type = Session.equals('layout', 'WheelLayout')
    || Session.equals('layout', 'ListLayout') ? 1 : 0;
  if (!c.firstRun) {
    itemSizeOld[type?0:1] = Tracker.nonreactive(function() {
      return Session.get('itemSize');
    });
    spacingOld[type?0:1] = Tracker.nonreactive(function() {
      return Session.get('spacing');
    });
  }
  Session.set('itemSize', itemSizeOld[type]);
  Session.set('spacing', spacingOld[type]);
});

Surfaces = new Meteor.Collection(null);

Meteor.startup(function() {
  for (var i=0; i < 60; i++)
    Surfaces.insert({
      name: '#'+(i+1),
      index: i,
      style: "background: hsl(" + (i * 360 / 60) + ", 100%, 50%)",
      show: Math.random() < 0.4
    });
});

Template.FlexScrollViewDemo.helpers({
  layout: function() {
    return Session.get('layout');
  },

  layoutOptions: function() {
    var out = {};
    _.each(layoutOptions[Session.get('layout')], function(option) {
      out[option] = Session.get(option);
    });
    return out;
  },

  direction: function() {
    return Session.get('direction');
  },

  surfaces: function() {
    return Surfaces.find({ show: true },
      { sort: { index: Session.get('order') } } );
  }
});

Session.setDefault('direction', 0);
Session.setDefault('layout', 'CollectionLayout');
Session.setDefault('order', 1);

Template.fsvHeader.events({
  'change input.picker': function(event) {
    Session.set(this.valueOf(), JSON.parse(event.currentTarget.value));
  },
  'click .surfaceAction': function(event) {
    var add = event.currentTarget.getAttribute('data-action') === 'add';
    var sids = _.pluck(Surfaces.find({show:!add}, {fields:{_id:1}}).fetch(), '_id');
    var id = sids[Math.floor((Math.random()-0.001) * sids.length)];
    Surfaces.update(id, { $set: { show: add }});
  },
  'change #layoutPicker': function(event) {
    Session.set('layout', event.currentTarget.value);
  },
  'change #directionPicker': function(event) {
    Session.set('direction', parseInt(event.currentTarget.value));
  },
  'change #orderPicker': function(event) {
    Session.set('order', parseInt(event.currentTarget.value));
  }
});

Template.fsvHeader.helpers({
  surfaceCount: function() {
    return Surfaces.find({show:true}).count();
  },

  layouts: _.keys(layoutOptions),

  directions: (function() {
    var out = [];
    for (key in famous.utilities.Utility.Direction)
      if (key !== 'Z') // ok this got a bit stupid :)
        out.push({key: key, value: famous.utilities.Utility.Direction[key]});
    return out;
  })(),

  layoutOptions: function() {
    return layoutOptions[ Session.get('layout') ];
  },
  layoutOptionValue: function() {
    return JSON.stringify( Session.get( this.valueOf() ) );
  },

  isLayoutSelected: function() {
    return Session.equals('layout', this.valueOf());
  },
  isDirectionSelected: function() {
    return Session.equals('direction', this.value);
  },
  isOrderDescending: function() {
    return Session.equals('order', -1);
  }
});
